core/src/sumeragi/main_loop.rs

```diff
@@ -216,101 +223,206 @@ impl<F: FaultInjection> SumeragiWithFault<F> {
-            );
-            *last_sent_transaction_gossip_time = Instant::now();
-        }
+        sumeragi.broadcast_packet(msg, current_topology);
     }
 }
 
-#[allow(clippy::expect_used)]
-fn commit_block<F>(
+fn commit_block<F: FaultInjection>(
     sumeragi: &SumeragiWithFault<F>,
-    block: VersionedCommittedBlock,
     state: &mut State,
-) where
-    F: FaultInjection,
-{
-    let block_hash = block.hash();
+    block: impl Into<VersionedCommittedBlock>,
+) {
+    let committed_block = block.into();
+    let block_hash = committed_block.hash();
 
     state
         .wsv
-        .apply(&block)
+        .apply(&committed_block)
         .expect("Failed to apply block on WSV. This is absolutely not acceptable.");
+    sumeragi.send_events(state.wsv.events_buffer.replace(Vec::new()));
+    sumeragi.send_events(&committed_block);
 
-    let events_buffer = state.wsv.events_buffer.replace(Vec::new());
-
-    // Update WSV copy that is public facing
     {
+        // Update WSV copy that is public facing
         let mut wsv = sumeragi.wsv.lock();
         *wsv = state.wsv.clone();
     }
 
-    if sumeragi.events_sender.receiver_count() > 0 {
-        for event in events_buffer.into_iter().chain(Vec::<Event>::from(&block)) {
-            trace!(?event);
-            sumeragi
-                .events_sender
-                .send(event)
-                .map_err(|e| warn!(%e, "Some events failed to be sent"))
-                .unwrap_or(0);
-            // Essentially log and ignore.
-        }
-    }
-
     state.previous_block_hash = state.latest_block_hash;
-    state.latest_block_height = block.header().height;
-    state.latest_block_hash = Some(block.hash());
-    state.latest_block_view_change_index = block.header().view_change_index;
+    state.latest_block_height = committed_block.header().height;
+    state.latest_block_hash = Some(committed_block.hash());
+    state.latest_block_view_change_index = committed_block.header().view_change_index;
+
+    let current_topology = &mut state.current_topology;
+    let role = current_topology.role(&sumeragi.peer_id);
+    current_topology.refresh_at_new_block(block_hash);
 
-    let previous_role = state.current_topology.role(&sumeragi.peer_id);
-    state.current_topology.refresh_at_new_block(block_hash);
     info!(
-        prev_peer_role = ?previous_role,
-        new_peer_role = ?state.current_topology.role(&sumeragi.peer_id),
-        new_block_height = %state.latest_block_height,
-        %block_hash,
-        "Committing block"
+        addr=%sumeragi.peer_id.address, %role,
+        block_height=%state.latest_block_height,
+        %block_hash, "Committing block"
     );
-    sumeragi.kura.store_block_blocking(block);
-    state.current_topology.update_network_topology(&state.wsv);
 
-    // Transaction Cache
-    cache_transaction(state, sumeragi)
+    sumeragi.kura.store_block_blocking(committed_block);
+    current_topology.update_network_topology(&state.wsv);
```

Comment:
I'm not sure if `update_network_topology` should be called after `refresh_at_new_block`? In all honesty, I think this should be joined into one function, because wsv can change peers only if block was committed. But the ordering might still be relevant

Reply:
@SamHSmith @Erigara

Reply:
As i see it peers only can be added through committing new block (`Register<Peer>/Unregister<Peer>`) so seems like `refresh_at_new_block` should include `update_network_topology`.

Reply:
I agree with that point, I'm just not sure which should be done first updating the network or refreshing. I would say first update(this is only done if there was a peer added/removed through ISI) then refresh the view

Reply:
I think it can be done at the same time by using topology `Builder`.

Reply:
got you `at_block` actually doesn't do anything per se. Will try to fix this before merging

Reply:
Order does not matter. `update_network_topology` only does something if the peers set in the wsv has changed.

Reply:
I've made this final change